fs: fix pre-aborted writeFile AbortSignal file leak

Linkgoron opened the pull request:
This fixes an issue in `promises.writeFile` where a file is opened, and not closed if the abort signal is aborted before/during the file was opened. 

A minor thing is that `fd.close` can still throw, I'm not sure if it's an issue or not.

In addition I also added validation for the AbortSignal, and a check where if the signal was _already_ aborted, the file won't be opened at all.

- [] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x]  commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

aduh95 commented:
Could you add a test? Also, do you know if it also happens on `fs.promises.readFile`?

Linkgoron commented:
> Could you add a test?
Sure

> do you know if it also happens on `fs.promises.readFile`?

It doesn't, however I could add the "optimization" the I added here (the file still opens, even if the signal is already aborted).

benjamingr reviewed (approved):
good catch

benjamingr commented:
(Please add a test)

Linkgoron commented:
@aduh95 @benjamingr  I've added tests. I'm not 100% happy about them, as it's hard for me to know if they'll work consistently... However, on my machine do work consistently on this branch, and fail consistently on master. I'd be happy to improve them if you have any pointers (The tests are essentially copy/pasted from my other fs PR, just changed for the promisified version)

RaisinTen commented on test/parallel/test-fs-promises-writefile.js:
```diff
@@ -25,9 +29,33 @@ async function doWriteWithCancel() {
+  assert.rejects(fsPromises.writeFile(otherDest2, buffer, { signal }), {
     name: 'AbortError'
-  });
+  }).then(common.mustCall(()=> {
```
```suggestion
  }).then(common.mustCall(() => {
```
To fix the linter errors.

RaisinTen commented on test/parallel/test-fs-promises-writefile.js:
```diff
@@ -25,9 +29,33 @@ async function doWriteWithCancel() {
+  controller.abort();
+  assert.rejects(fsPromises.writeFile(otherDest3, buffer, { signal }), {
+    name: 'AbortError'
+  }).then(common.mustCall(()=> {
```
```suggestion
  }).then(common.mustCall(() => {
```

RaisinTen commented on test/parallel/test-fs-promises-writefile.js:
```diff
@@ -25,9 +29,33 @@ async function doWriteWithCancel() {
+  controller.abort();
+  assert.rejects(fsPromises.writeFile(otherDest4, buffer, { signal }), {
+    name: 'AbortError'
+  }).then(common.mustCall(()=> {
```
```suggestion
  }).then(common.mustCall(() => {
```

aduh95 commented on lib/internal/fs/promises.js:
```diff
@@ -668,11 +669,17 @@ async function writeFile(path, data, options) {
+    throw lazyDOMException('The operation was aborted', 'AbortError');
+  }
+
   const fd = await open(path, flag, options.mode);
   if (options.signal?.aborted) {
+    await fd.close();
     throw lazyDOMException('The operation was aborted', 'AbortError');
   }
   return PromisePrototypeFinally(writeFileHandle(fd, data), fd.close);
```
Shouldn't we throw the DOMException right? In case `fd.close()` is rejected or takes a long time to resolve, it would create a weird behavior.
```suggestion
  const fd = await open(path, flag, options.mode);
  try {
    if (options.signal?.aborted) {
      throw lazyDOMException('The operation was aborted', 'AbortError');
    }
    return await writeFileHandle(fd, data);
  } finally {
    await fd.close();
  }
```

Linkgoron replied:
I don't mind changing to this, I just don't think that this would behave differently (as the finally happens before the promise resolves, and if you throw in the finally it would still throw).

Linkgoron replied:
I've changed the code to match the above

aduh95 reviewed (commented):
If making a robust test is too tricky, I think we should land this without one.

Linkgoron commented:
removed test changes

benjamingr reviewed (approved):
lgtm without the test due to the circumstances.

aduh95 commented on lib/internal/fs/promises.js:
```diff
@@ -668,11 +669,17 @@ async function writeFile(path, data, options) {
+  }
+
   const fd = await open(path, flag, options.mode);
   if (options.signal?.aborted) {
+    await fd.close();
     throw lazyDOMException('The operation was aborted', 'AbortError');
   }
   return PromisePrototypeFinally(writeFileHandle(fd, data), fd.close);
```
What doo you think of this:
```suggestion
  return PromisePrototypeFinally(
    options.signal?.aborted ?
      PromiseReject(lazyDOMException('The operation was aborted', 'AbortError')) :
      writeFileHandle(fd, data),
    fd.close
  );
```

Linkgoron replied:
I think that the other way you suggested is better (with the try/finally)

aduh95 commented on lib/internal/fs/promises.js:
```diff
@@ -692,6 +702,10 @@ async function readFile(path, options) {
+    throw lazyDOMException('The operation was aborted', 'AbortError');
+  }
+
   const fd = await open(path, flag, 0o666);
   return PromisePrototypeFinally(readFileHandle(fd, options), fd.close);
```
What if the signal is aborted while we wait for `open`? Shouldn't we have the same `try { } finally { }` pattern as in `writeFile`?

Linkgoron replied:
readFileHandle already throws

aduh95 replied:
So does `writeFileHandle`, maybe we should remove the change above (line 678..).

Linkgoron replied:
While true, there's a minor difference here which why I changed it (as that was my first instinct as well). `writeFileHandle` returns without error if the amount of data to be write is `0` bits. I assumed that that's why the function was written in the original way (with the abort signal check before the write). If this change is immaterial, I think that removing the changes of line 678 is the best solution to this PR.

Linkgoron replied:
Oh, re-reading it I assume that it was written this way to avoid allocating the buffer which might be large...

aduh95 replied:
Yeah, removing it is probably the right move here

aduh95 replied:
> Oh, re-reading it I assume that it was written this way to avoid allocating the buffer which might be large...

OK, so maybe we can add a comment explaining why it is done differently.

Linkgoron replied:
I removed the signal check, and left the original `PromisePrototypeFinally`. From what I understand, the `Uint8Array` constructor shouldn't allocate much because it only offers a "view" into the buffer that it receives, so unless I'm misunderstanding something - it shouldn't really matter.

Linkgoron replied:
@aduh95 I missed that `return PromisePrototypeFinally(readFileHandle(fd, options), fd.close);` didn't even pass the signal into the function. I've pushed a fix...

aduh95 commented:
Landed in c4180b74c5ba
